zips/zip-0302.rst

```diff
@@ -64,6 +67,57 @@ contents of a memo. It does not define consensus requirements.
   assumption about the memo. This value was used ambiguously in the past by private
   agreement; applications SHOULD prefer ``0xFF`` which is unambiguously for this purpose.
 
+Structured memos
+----------------
+
+A memo can be used to store structured data by setting its first byte to a value of
+``0xF7``. The remaining bytes contain a set of "memo parts", encoded as concatenated
+Type-Length-Value tuples, followed by fixed padding (if any memo bytes remain after the
+TLV tuples).
+
+Multiple memo parts with the same type are not permitted in structured memos.
+
+Parsing a structured memo
+`````````````````````````
+
+After the leading ``0xF7`` byte, readers parse the remaining memo bytes as follows:
+
+- While there are remaining unparsed memo bytes:
+
+  - Parse a ``compactSize`` to obtain ``partType``. If there are insufficient bytes to
+    parse a ``compactSize``, return an error.
+  - If ``partType == 0``:
+
+    - This is a marker type indicating that there are no more TLV tuples. If any of the
+      remaining memo bytes are not ``0x00``, return an error.
+
+  - Else if ``partType`` has already been parsed in an earlier memo part, return an error.
+  - Otherwise:
+
+    - Parse a ``compactSize`` to obtain ``partLength``. If there are insufficient bytes to
+      parse a ``compactSize``, return an error.
+    - Parse the next ``partLength`` bytes to obtain ``partValue``. If ``partLength`` would
+      require the reader to read past the end of the memo bytes, return an error.
+    - Store ``(partType, partValue)`` and continue parsing.
+
+If an error occurs, readers MUST treat the entire memo as invalid (ignoring any
+``MemoPart`` that was successfully parsed prior to the error).
+
+Defined memo part types
+```````````````````````
+
+The following memo part types (and corresponding value encodings) are defined:
+
+- ``0xA0``: a UTF-8 string.
```

Comment:
Write part types as decimal integers, so we don't have any confusion between the part types and their encodings.